astropy/io/fits/hdu/hdulist.py

```diff
@@ -661,7 +661,7 @@ def index_of(self, key):
             _ver = None
 
         if not isinstance(_key, string_types):
-            raise TypeError(
+            raise KeyError(
                 '%s indices must be integers, extension names as strings, '
```

Comment:
Minor nitpicking, but could you change these strings here and a few other places in the file to use the ``.format()`` formatting instead of ``%``